File: src_code/0021-merge-two-sorted-lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if(l1==null){
            return l2;
        }
        if(l2==null){
            return l1;
        }
        ListNode dummy = new ListNode(0);
        ListNode res = dummy;
        while(l1!=null&&l2!=null){
            if(l1.val<l2.val){
                dummy.next = new ListNode(l1.val);
                l1 = l1.next;
                dummy = dummy.next;
            }else if(l2.val<l1.val){
                dummy.next = new ListNode(l2.val);
                l2 = l2.next;
                dummy = dummy.next;
            }else{
                dummy.next = new ListNode(l1.val);
                dummy = dummy.next;
                dummy.next = new ListNode(l2.val);
                l1 = l1.next;
                l2 = l2.next;
                dummy = dummy.next;
                
            }
        }
        
        while(l1!=null){
            dummy.next = new ListNode(l1.val);
            dummy = dummy.next;
            l1 = l1.next;
        }
        while(l2!=null){
            dummy.next = new ListNode(l2.val);
            dummy = dummy.next;
            l2 = l2.next;
        }
        return res.next;
        
    }
}

File: src_code/0021-merge-two-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        Merge two sorted linked lists.
        
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        # Create a dummy node as the starting point of the merged list
        dummy = ListNode(-1)
        current = dummy  # Pointer to the current node in the merged list

        while list1 and list2:
            # Compare the values of the two lists' nodes
            if list1.val < list2.val:
                current.next = list1
                list1 = list1.next
            else:
                current.next = list2
                list2 = list2.next
            current = current.next

        # If one of the lists is not empty, append the remaining part to the merged list
        if list1:
            current.next = list1
        if list2:
            current.next = list2

        return dummy.next  # The merged list starts from the next node of the dummy
        
